op-node/rollup/derive/batches_espresso_test.go

```diff
@@ -0,0 +1,614 @@
+		},
+		{
+			Name:       "sequencer time drift on same epoch with empty txs and but in-sight epoch that invalidates it",
+			L1Blocks:   []eth.L1BlockRef{l1A, l1B, l1C},
+			L2SafeHead: l2A3,
+			Batch: BatchWithL1InclusionBlock{
+				L1InclusionBlock: l1C,
+				Batch: &BatchData{BatchV2{
+					BatchV1: BatchV1{ // we build l2A4, which has a timestamp of 2*4 = 8 higher than l2A0
+						ParentHash:   l2A4.ParentHash,
+						EpochNum:     rollup.Epoch(l2A4.L1Origin.Number),
+						EpochHash:    l2A4.L1Origin.Hash,
+						Timestamp:    l2A4.Time,
+						Transactions: nil,
+					},
+					Justification: nil,
+				}},
+			},
+			Expected: BatchDrop, // dropped because it could have advanced the epoch to B
+		},
+		{
+			Name:       "empty tx included",
+			L1Blocks:   []eth.L1BlockRef{l1A, l1B},
+			L2SafeHead: l2A0,
+			Batch: BatchWithL1InclusionBlock{
+				L1InclusionBlock: l1B,
+				Batch: &BatchData{BatchV2{
+					BatchV1: BatchV1{
+						ParentHash: l2A1.ParentHash,
+						EpochNum:   rollup.Epoch(l2A1.L1Origin.Number),
+						EpochHash:  l2A1.L1Origin.Hash,
+						Timestamp:  l2A1.Time,
+						Transactions: []hexutil.Bytes{
+							[]byte{}, // empty tx data
+						},
+					},
+					Justification: nil,
+				}},
+			},
+			Expected: BatchDrop,
+		},
+		{
+			Name:       "deposit tx included",
+			L1Blocks:   []eth.L1BlockRef{l1A, l1B},
+			L2SafeHead: l2A0,
+			Batch: BatchWithL1InclusionBlock{
+				L1InclusionBlock: l1B,
+				Batch: &BatchData{BatchV2{
+					BatchV1: BatchV1{
+						ParentHash: l2A1.ParentHash,
+						EpochNum:   rollup.Epoch(l2A1.L1Origin.Number),
+						EpochHash:  l2A1.L1Origin.Hash,
+						Timestamp:  l2A1.Time,
+						Transactions: []hexutil.Bytes{
+							[]byte{types.DepositTxType, 0}, // piece of data alike to a deposit
+						},
+					},
+					Justification: nil,
+				}},
+			},
+			Expected: BatchDrop,
+		},
+		{
+			Name:       "valid batch where hotshot transactions fall within the window",
+			L1Blocks:   []eth.L1BlockRef{l1A, l1B, l1C},
+			L2SafeHead: l2A3,
+			Batch: BatchWithL1InclusionBlock{
+				L1InclusionBlock: l1A,
+				Batch: &BatchData{BatchV2{
+					BatchV1: BatchV1{
+						ParentHash: l2B0.ParentHash,
+						EpochNum:   rollup.Epoch(l2B0.L1Origin.Number),
+						EpochHash:  l2B0.L1Origin.Hash,
+						Timestamp:  l2B0.Time,
+						Transactions: []hexutil.Bytes{
+							[]byte{0x02, 0x42, 0x13, 0x37},
+							[]byte{0x02, 0xde, 0xad, 0xbe, 0xef},
+						},
+					},
+					Justification: &eth.L2BatchJustification{
+						FirstBlock: espresso.Header{
+							Timestamp: l2B0.Time,
+						},
+						PrevBatchLastBlock: espresso.Header{
+							Timestamp: l2B0.Time - 1,
+						},
+						FirstBlockNumber: 1,
+						Payload: &eth.L2BatchPayloadJustification{
+							LastBlock: espresso.Header{
+								Timestamp: l2B0.Time + conf.BlockTime - 1,
+							},
+							NextBatchFirstBlock: espresso.Header{
+								Timestamp: l2B0.Time + conf.BlockTime,
+							},
```

Comment:
This seems like it should fail, since we don't have `NmtProofs` with length at least 2, and yet `FirstBlock` and `LastBlock` are both included in the range. We should add the `NmtProofs` but also something might be wrong with the check blocks function if this is passing?

Reply:
We should also add a test similar to this one, but with 3 NmtProofs, implying there was one additional block in the window between FirstBlock and LastBlock

Reply:
I can add this test in, but I'm not sure how useful it is right now because the header/NMT validation is mocked. I can at least add a sanity check that there are least 1 or 2 NMTProofs depending on whether FirstBlock != LastBlock

Reply:
I don't think this needs to be an explicit sanity check (not that it would be bad to add one). Don't the `NmtProofs` determine the block number that we use when verifying the commitment for `LastBlock`? So if there are the wrong number of NmtProofs, the commitment validation should fail? Actually I'm confused why it's not segfaulting right now since we use `nil` for the `HotShotContractProvider`

Reply:
Yes that's true - I think maybe its because this code is only reached in the unit test where we provide a mock HotShotContractProvider?